bundles/org.openhab.binding.pjlinkdevice/src/main/java/org/openhab/binding/pjlinkdevice/internal/discovery/AbstractDiscoveryParticipant.java

```diff
@@ -0,0 +1,133 @@
+/**
+ * Copyright (c) 2010-2019 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.pjlinkdevice.internal.discovery;
+
+import java.net.InetAddress;
+import java.net.InterfaceAddress;
+import java.net.NetworkInterface;
+import java.net.SocketException;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
+import org.eclipse.jdt.annotation.NonNull;
+import org.eclipse.smarthome.config.discovery.AbstractDiscoveryService;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.eclipse.smarthome.core.thing.ThingUID;
+import org.openhab.binding.pjlinkdevice.internal.PJLinkDeviceBindingConstants;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+
+/**
+ * Discovery of PJLink devices. Checks IP addresses in parallel processing.
+ * 
+ * Generating IP addresses and checking them is done by the subclasses implementing
+ * {@link AbstractDiscoveryParticipant#generateAddressesToScan} and {@link AbstractDiscoveryParticipant#checkAddress}
+ * 
+ * @author Nils Schnabel - Initial contribution
+ */
+@NonNullByDefault
+public abstract class AbstractDiscoveryParticipant extends AbstractDiscoveryService {
+  protected final Logger logger = LoggerFactory.getLogger(DiscoveryParticipantClass1.class);
+  private Integer scannedIPcount = 0;
+  private @Nullable ExecutorService executorService = null;
+
+  public AbstractDiscoveryParticipant(Set<@NonNull ThingTypeUID> supportedThingTypes, int timeout,
+      boolean backgroundDiscoveryEnabledByDefault) throws IllegalArgumentException {
+    super(supportedThingTypes, timeout, backgroundDiscoveryEnabledByDefault);
+  }
+
+  protected ExecutorService getExecutorService() {
+    ExecutorService executorService = this.executorService;
```

Comment:
```suggestion
    final ExecutorService executorService = this.executorService;
```

Reply:
This won't work as I'm setting it to a new instance two lines later.